Inject Redis via named constructor

martinssipenko opened the pull request:
This PR adds ability to create Redis storage class by injecting existing Redis connection.

Clsoes #7

martinssipenko committed:
Inject Redis via named constructor

NoelDavies commented on src/Prometheus/Storage/Redis.php:
```diff
@@ -51,6 +56,15 @@ public function __construct(array $options = [])
         $this->redis = new \Redis();
     }
 
+    public static function withRedis(\Redis $redis): self
```
`fromExistingConnection()` may be a better named constructor

martinssipenko committed:
change method name

martinssipenko commented:
If you are happy with this way of doing it I'll add tests.

NoelDavies commented:
> If you are happy with this way of doing it I'll add tests.

I personally think this works! I'd suggest in the future you write tests first, allowing you to design your API first.

But looks good, just add tests. Nice addition to the project 👍

martinssipenko committed:
Added withExistingConnection test

martinssipenko committed:
Merge branch 'master' into inject-redis

* master:
  fix(composer): add replace rule
  Add Installation instructions

martinssipenko committed:
Fix for redis prefix support and respective tests

martinssipenko committed:
removed unneeded code

martinssipenko commented:
@NoelDavies while trying to integrate my forked branch into my project I noticed that when I inject Redis connection with set prefix, metrics stop working. After poking around for a bit I discovered that PhpRedis adds a prefix to `KEYS` only, and not `ARGS`. As a result, before my changes, if prefix used lead to redis commands being prefixed too, which lead to errors.

I added tests to cover this case and afterwards changed the `KEYS` and `ARGS` order.  Now this works with and without redis prefix, all tests pass.

I also added an extra check in `Redis::fromExistingConnection()` method so that if connection is not established exception is thrown.

I hope this can be released soon.

piotrooo commented:
You know, in this implementation you are close only for one of the Redis implementation and it's breaks an inverse of control pattern.

This implementation should be a default Redis implementation, which implement some RedisClientIntreface, that allows easly extend Redis client implementation by the dev which uses diffrent Redis client.

martinssipenko commented:
@piotrooo I can't disagree, however this was not the intention of this pull request. Given implementation already is based on PhpRedis `\Redis` class this PR only adds ability to inject same implementation from outside. Perhaps this could be discussed and changed in new major version.

piotrooo commented:
@martinssipenko sounds reasonably. Should be discussed and reimplemented - IMHO.

martinssipenko committed:
Merge branch 'master' into inject-redis

* master:
  Fix invalid json (typo)

martinssipenko commented:
@NoelDavies any updates on this?

NoelDavies commented on src/Prometheus/Storage/Redis.php:
```diff
@@ -99,6 +117,10 @@ function (array $metric) {
      */
     private function openConnection(): void
     {
+        if ($this->connectionInitialized) {
```
This should be a strict comparison

NoelDavies commented on src/Prometheus/Storage/Redis.php:
```diff
@@ -51,6 +56,19 @@ public function __construct(array $options = [])
 
+    public static function fromExistingConnection(\Redis $redis): self
+    {
+        if (false === $redis->isConnected()) {
```
we're using non-yoda style comparisons in the rest of the codebase. Please change this to be `actual === expected`

NoelDavies replied:
This still needs changing

martinssipenko replied:
It is fixed, you are looking at outdated change.

martinssipenko committed:
Update according to feedback

martinssipenko commented:
@NoelDavies any updates on this?

martinssipenko commented:
@NoelDavies I think you were looking at outdated version of code, all feedback that was given has been implemented - https://github.com/endclothing/prometheus_client_php/pull/8/files

NoelDavies commented:
Can we rebase from master to get the CI changes ?

martinssipenko committed:
Merge branch 'master' into inject-redis

* master:
  Add support for prometheus/pushgateway:>=0.10 (#17)
  Introduce Continuous Integration (#20)

martinssipenko commented:
Done, this branch is now up to date with upstream master. However, the CI is still failing because of incompatible Guzzle version, that I fixed in #21. If you want I can fix the CI here or in a different branch forked from current master.

martinssipenko committed:
Merge branch 'master' into inject-redis

* master:
  Add code style check into CI pipeline (#21)

NoelDavies commented:
I merged #21 now, it should run the CI fine now?

martinssipenko committed:
phpcbf

martinssipenko commented:
Rebased again, CI should be green now, also other PRs should be rebased and be green before merging (you can require that CI passes before merge is allowed in githubs branch protection settings).
